Bluetooth: Host: Remove blocking behaviour from `gatt_req_alloc`

alwa-nordic opened the pull request:
bt_gatt_write would block in call to gatt_req_alloc if there were no
free req objects. `bt_gatt_write` is documented as async, so it should
not block.  This also affects all other `bt_gatt_*` functions that
allocate a req object.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/39624 where
if `bt_gatt_write` is called from BT RX thread (as can happen if it is
called from a bluetooth callback), the BT RX thread can be blocked and
prevented from processing the request responses and unblocking itself.
This was the cause of a soft 30s deadlock until gatt_req_alloc timeouts.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

jhedberg commented on include/bluetooth/gatt.h:
```diff
@@ -1465,6 +1465,15 @@ struct bt_gatt_write_params {
  *  @param params Write parameters.
  *
  *  @return 0 in case of success or negative value in case of error.
+ *
+ *  @retval 0
+ *      Successfully queued request.
+ *      Will call @p params->func on resolution.
+ *
+ *  @retval -ENOMEM
+ *      Too many pending ATT requests.
+ *      Allow a pending request to resolve before retrying.
+ *      Relevant limit is @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
```
This constitutes in some ways API breakage (in the sense that some application might have depended on the blocking behaviour for correct functionality). Would it make more sense to introduce a `bt_gett_write_async()` along side of the old API? If we don't do that we should follow the API change process and clearly document this in the release notes for Zephyr 3.0.

alwa-nordic replied:
I see your point. But just above in the function doc we have:

>  This procedure is asynchronous therefore the parameters need to remains valid while it is active.

So there was a mismatch between the documented API and behavior. This change makes the function adhere to its API. Given the function doc above, `bt_gatt_write_async` would be confusing.

Does this change your mind on this, @jhedberg?

Of course we should consider if we are breaking someones code. The blocking behavior could easily be relied on as flow control.

We could instead ratify the blocking behavior in the API. It may be possible to give a more informative LOG_ERR about the timeout. We could detect that we are on BT RX thread and report it as a deadlock. But what if we are in a more complex situation involving a transitive blocking where BT RX thread waits on different thread that is stuck in `bt_gatt_write`. Maybe we could detect a blocked BT RX thread on timeout and log it as the probable reason for the timeout. But it gets needlessly complex.

I prefer removing the blocking to remove any concerns of a deadlock. But I know how unpopular breaking other peoples code for the sake of correctness is. I would not be against scheduling this change for a major release.

jhedberg replied:
> Does this change your mind on this, @jhedberg?

Yes, that's a pretty good point. However, I still think it would be good to get maximal visibility on this change by having an entry in the release notes for it.

joerchan replied:
This seems to be the same issue that we had with conn_tx_alloc.
Which we solved like this:
```
static struct bt_conn_tx *conn_tx_alloc(void)
{
	/* The TX context always get freed in the system workqueue,
	 * so if we're in the same workqueue but there are no immediate
	 * contexts available, there's no chance we'll get one by waiting.
	 */
	if (k_current_get() == &k_sys_work_q.thread) {
		return k_fifo_get(&free_tx, K_NO_WAIT);
	}
```
This is also documented in the API headers.

Not saying any solution is better than the other, but it gets confusing when it it is handled in multiple different ways.

alwa-nordic replied:
Good idea. `bt_gatt_write` is not the only public function that has changed. I think we should mention them all in the release notes. I'll have a look and find them. Should it go into doc/releases/release-notes-3.0.rst?

Thalley replied:
> This procedure is asynchronous therefore the parameters need to remains valid while it is active.

@alwa-nordic This line got me thinking: When you read this, it sounds like procedure == function to you, but I am not sure I agree with that. The function may still be blocking, even if it an asynchronous procedure, so I don't agree that there is a mismatch between the documented API and the behavior. 

In any case, the blocking behavior of the function have caused significant amounts of issues when the receiver of callbacks in the RX thread attempts to send additional requests, so I am not against this change at all.

It will be different behavior from the POSIX `send()` function though. Not sure how badly Zephyr tries to be similar to POSIX.

carlescufi commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2018,8 +2018,8 @@ static struct bt_att_req *gatt_req_alloc(bt_att_func_t func, void *params,
-	/* Allocate new request */
-	req = bt_att_req_alloc(BT_ATT_TIMEOUT);
+	/* This function is called from asynchronous bt_gatt functions and shall not block/wait. */
+	req = bt_att_req_alloc(K_NO_WAIT);
```
```
	if (k_current_get() == &bt_rx_thread) {
		return bt_att_req_alloc(K_NO_WAIT);
	}
```

carlescufi replied:
And then add a comment similar to https://github.com/zephyrproject-rtos/zephyr/blob/main/include/bluetooth/gatt.h#L985

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1208,10 +1208,30 @@ struct bt_gatt_exchange_params {
  *  @note Shall only be used once per connection.
  *
+ *  The Response comes in callback @p params->func. The callback is run from
+ *  the Bluetooth event context. @p *params must remain valid until start of
```
"Bluetooth event context" is not described or mentioned anywhere else. I suppose you mean the RX thread?

Thalley replied:
```suggestion
 *  the Bluetooth event context. @p params must remain valid until start of
```

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1208,10 +1208,30 @@ struct bt_gatt_exchange_params {
  *
  *  @return 0 in case of success or negative value in case of error.
+ *
+ *  @retval 0
+ *      Successfully queued request.
+ *      Will call @p params->func on resolution.
+ *
+ *  @retval -ENOMEM
+ *      ATT request queue is full and blocking would cause deadlock.
+ *      Allow a pending request to resolve before retrying, or call
+ *      this function outside Bluetooth event context to get blocking
+ *      behavior.
+ *      Queue size is controlled by @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
```
Uh, I like the use of `@retval`. That's something I'd like to see more of in the Bluetooth headers

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1340,13 +1360,31 @@ struct bt_gatt_discover_params {
- *  @note This procedure is asynchronous therefore the parameters need to
- *        remains valid while it is active.
+ *  The Response comes in callback @p params->func. The callback is run from
+ *  the Bluetooth event context. @p *params must remain valid until start of
```
```suggestion
 *  the Bluetooth event context. @p params must remain valid until start of
```

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1208,10 +1208,30 @@ struct bt_gatt_exchange_params {
+ *  callback.
+ *
+ *  This function will block while the ATT request queue is full, except when
+ *  called from Bluetooth event context. When called from Bluetooth context,
+ *  this function will instead instead return `-ENOMEM` if it would block to
+ *  avoid a deadlock.
```
It's not bad to have, but it does seem a bit superfluous to describe the return value both here and in the `@retval`

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1425,13 +1463,31 @@ struct bt_gatt_read_params {
- *  @note This procedure is asynchronous therefore the parameters need to
- *        remains valid while it is active.
+ *  The Response comes in callback @p params->func. The callback is run from
+ *  the Bluetooth event context. @p *params must remain valid until start of
```
```suggestion
 *  the Bluetooth event context. @p params must remain valid until start of
```

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4607,45 +4640,49 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
+	if (!sub_has_other_subscriptions_to_same_attribute(sub, params)) {
+		int gatt_write_err;
+
+		params->value = 0x0000;
```
Suggest to create a 
```
#define BT_GATT_CCC_UNSUB			0x0000
```
and use that here

alwa-nordic replied:
I chose not to do this because it seems to me that `BT_GATT_CCC_*` is intended for `bt_gatt_subscribe_params.value`. And it makes little sense to allow BT_GATT_CCC_UNSUB there.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4587,13 +4592,41 @@ int bt_gatt_resubscribe(uint8_t id, const bt_addr_le_t *peer,
 	struct bt_gatt_subscribe_params *tmp, *next;
-	bool has_subscription = false, found = false;
 	sys_snode_t *prev = NULL;
+	bool wait_for_ccc_write_response = false;
```
Not crazy about this name, as we are actually not really waiting for anything.

jhedberg replied:
How about calling this `other_sub' and not pre-initializing it, rather pass the return value from `has_other_sub()` to it?

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3363,8 +3363,29 @@ void hci_event_prio(struct net_buf *buf)
 }
 
+static k_tid_t bt_recv_thread_id;
+bool bt_recv_is_blocked_by_current_thread(void)
```
Just call this `bt_is_recv_thread()'

Thalley reviewed (changes requested):
Minor comments, otherwise the change looks good

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4587,13 +4592,41 @@ int bt_gatt_resubscribe(uint8_t id, const bt_addr_le_t *peer,
+	return false;
+}
+
+static bool sub_has_other_subscriptions_to_same_attribute(struct gatt_sub *sub,
```
You really seem to like long symbol names 😉 

How about just `has_other_sub()`?

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3363,8 +3363,29 @@ void hci_event_prio(struct net_buf *buf)
+
 int bt_recv(struct net_buf *buf)
 {
+	save_bt_recv_thread_id();
```
inline the function contents directly in `bt_recv`

alwa-nordic committed:
Bluetooth: Host: Handle req alloc failure in bt_gatt_unsubscribe

`gatt_write_ccc` may fail if it is unable to allocate a `bt_att_req`.
This change adds handling of this case to `bt_gatt_unsubscribe`.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Host: Detect deadlock in bt_att_req_alloc

`gatt_req_alloc` will wait until a `req` is free (or until timeout).
`req`s are freed on the BT RX thread in calls into bt_att_recv.

When `gatt_req_alloc` called on the BT RX thread itself when there are
no free `req`s, it will block the BT RX thread and deadlock. The
deadlock lasts until timeout.

This change detects this condition and returns the failure early.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/39624 where
if `bt_gatt_write` is called from BT RX thread (as can happen if it is
called from a bluetooth callback), the BT RX thread can be blocked and
prevented from processing the request responses and unblocking itself.
This was the cause of a soft 30s deadlock until gatt_req_alloc timeouts.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Host: Update API docs on GATT request failures

Document the deadlock avoidance in ATT request queueing.

Add more precise description of the callback and how long the argument
must be kept valid for.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic commented:
I had a bug in my previous change set. I forgot to wrap the ccc write with `if (!has_subscription)`. Fixed now.

alwa-nordic commented:
I tried to de-duplicate the documentation using `@snippet{doc} this my_snippet`. But it did not work well. The delimited snippet (surrounded by `[my_snippet]`) has to be in a Doxygen block since it's not C code. But we have a rule that comments must start with a `*`. But the snippet is captured verbatim and `*` is not stripped even if the snippet is inside a Doxygen block. The `*` parsed as markdown and ends up producing a bullet list.

carlescufi commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4608,19 +4607,10 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
 	/* Lookup existing subscriptions */
-	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&sub->list, tmp, next, node) {
-		/* Remove subscription */
+	SYS_SLIST_FOR_EACH_CONTAINER(&sub->list, tmp, node) {
```
instead of `continue` you could have `else if()`

carlescufi reviewed (changes requested):
Need to check with @gmarull whether there is a better alternative to the duplication in doxygen.

gmarull commented on include/bluetooth/gatt.h:
```diff
@@ -1340,13 +1356,27 @@ struct bt_gatt_discover_params {
  *  @param conn Connection object.
  *  @param params Discover parameters.
  *
  *  @return 0 in case of success or negative value in case of error.
+ *
+ *  @retval 0
+ *      Successfully queued request.
+ *      Will call @p params->func on resolution.
+ *
+ *  @retval -ENOMEM
+ *      ATT request queue is full and blocking would cause deadlock.
+ *      Allow a pending request to resolve before retrying, or call this
+ *      function outside the BT RX thread to get blocking behavior.
+ *      Queue size is controlled by @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
```
this documentation is now inconsistent, looks like first `@return` is now redundant?

better use this format, for consistency:
```c
 *  @retval -ENOMEM ATT request queue is full and blocking would cause deadlock.
 *  Allow a pending request to resolve before retrying, or call this function outside the
 *  BT RX thread to get blocking behavior. Queue size is controlled by 
 *  @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
```

carlescufi commented on include/bluetooth/gatt.h:
```diff
@@ -1340,13 +1356,27 @@ struct bt_gatt_discover_params {
  *
- *  @note This procedure is asynchronous therefore the parameters need to
- *        remains valid while it is active.
+ *  The Response comes in callback @p params->func. The callback is run from
```
We could potentially use something like pinctrl here, its own "named" section: https://docs.zephyrproject.org/latest/doxygen/html/group__pinctrl__interface.html#ga46d7997af7e9231b431c79af34288d50

carlescufi reviewed (changes requested):
Doxygen needs rework

jsiverskog commented:
@alwa-nordic : what's the plan for this mr?

carlescufi commented on include/bluetooth/gatt.h:
```diff
@@ -1216,10 +1216,24 @@ struct bt_gatt_exchange_params {
  *  @param conn Connection object.
  *  @param params Exchange MTU parameters.
  *
  *  @return 0 in case of success or negative value in case of error.
```
remove this line
